src/Runner.Worker/Container/ContainerHooks/ContainerHookManager.cs

```diff
@@ -0,0 +1,308 @@
+            var handler = handlerFactory.Create(
+                            context,
+                            null,
+                            stepHost,
+                            new ScriptActionExecutionData(),
+                            inputs,
+                            environment: new Dictionary<string, string>(VarUtil.EnvironmentVariableKeyComparer),
+                            context.Global.Variables,
+                            actionDirectory: scriptDirectory,
+                            localActionContainerSetupSteps: null) as ScriptHandler;
+            handler.PrepareExecution(stage);
+
+            IOUtil.CreateEmptyFile(input.ResponseFile);
+            await handler.RunAsync(stage);
+            if (handler.ExecutionContext.Result == TaskResult.Failed)
+            {
+                throw new Exception($"The hook script at '{HookIndexPath}' running command '{input.Command}' did not execute successfully."); // TODO: better exception
+            }
+            var response = GetResponse<T>(input);
+            return response;
+        }
+
+        private string GenerateResponsePath()
+        {
+            return Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Temp), ResponseFolderName, $"{Guid.NewGuid()}.json");
+        }
+
+        private string GetPrependPath(IExecutionContext context)
+        {
+            return string.Join(Path.PathSeparator.ToString(), context.Global.PrependPath.Reverse<string>()); ;
+        }
+
+        private void ValidateHookExecutable()
+        {
+            if (!string.IsNullOrEmpty(HookIndexPath) && !File.Exists(HookIndexPath))
+            {
+                throw new Exception($"File not found at '{HookIndexPath}'. Set {Constants.Hooks.ContainerHooksPath} to the path of an existing file.");
+            }
+
+            var supportedHookExtensions = new string[] { ".js", ".sh", ".ps1" };
+            if (!supportedHookExtensions.Any(extension => HookIndexPath.EndsWith(extension)))
+            {
+                throw new Exception($"Invalid file extension at '{HookIndexPath}'. {Constants.Hooks.ContainerHooksPath} must be a path to a file with one of the following extensions: {string.Join(", ", supportedHookExtensions)}");
+            }
+        }
+
+        private T GetResponse<T>(HookInput input) where T : HookResponse
+        {
+            if (!File.Exists(input.ResponseFile))
+            {
+                Trace.Info($"Response file for the hook script at '{HookIndexPath}' running command '{input.Command}' not found.");
+                if (input.Command == HookCommand.PrepareJob)
+                {
+                    throw new Exception($"Response file is required but not found for the hook script at '{HookIndexPath}' running command '{input.Command}'");
+                }
+                return null;
+            }
+
+            T response = IOUtil.LoadObject<T>(input.ResponseFile);
```

Comment:
do we need some protection or error handling for invalid JSON from the hook script?

Reply:
We do catch them in the hook methods individually, I think that should be brought down to ExecuteHookScript